modules/ding_place2book/ding_place2book.module

```diff
@@ -681,19 +686,31 @@ function ding_place2book_get_event_state($event_maker_id, $event_id) {
         }
 
         $now = time();
-        foreach ($sale_periods as $sale_period) {
-          if ($now >= $sale_period['begin'] && $now <= $sale_period['end']) {
-            // We have found an open sales period. No need to continue iterating
-            // the ticket types.
-            $ticket_link_status = 'open';
-            break;
-          }
-          elseif ($now < $sale_period['begin']) {
-            // We haven't found any open sales period yet,q but at least there's
-            // an upcoming, so set that status for now.
-            $ticket_link_status = 'upcoming';
-          }
+
+        // Filter out periods which are already over.
+        $sale_periods = array_filter($sale_periods, function ($p) use ($now) {
+          return $p['end'] >= $now;
+        });
+
+        // Find the closest sale period.
+        $begin_values = array_column($sale_periods, 'begin');
+        $min_period_begin = min($begin_values);
+        $min_period = array_filter($sale_periods, function ($sale_period) use ($min_period_begin) {
+          return ($sale_period['begin']) == $min_period_begin;
+        });
+        $min_period = array_shift($min_period);
+
+        if ($now >= $min_period['begin'] && $now <= $min_period['end']) {
```

Comment:
How will this behave if:

1. All sales periods are over and there are no $min_period?
2. The event is published and non-passive, but somehow there's no sales periods on the ticket and the event and hence also no $min_peroid.

Also a bit nitpicking here: wouldn't it be more simple to just check `$now < $min_period['begin']` first and then 'upcoming' otherwise just 'open', since you have already verified that all the sales periods left are not over?

Reply:
Checked one more time.

> All sales periods are over.

If such situation happens when the status will be set to "event-over" or "closed".

> The event is published and non-passive, but somehow there's no sales periods on the ticket.

In such case the request to P2B will not return anything and this simply will not proceed, because once the module is enabled and configured - it will force the authors to define at least one price even the event is free.

Reply:
> If such situation happens when the status will be set to "event-over" or "closed".

No it won't. Just because all the sales periods are over doesn't mean the event is over or complete. You will get a notice in these cases,

> In such case the request to P2B will not return anything and this simply will not proceed, Once the module is enabled and configured - it will force the authors to define at least one price even the event is free.

I'm not sure what you mean it will not return anything(?), but it's possible to delete a price in place2book GUI after creation and in these situations you will also get a notice.

Reply:
@bozeak I just tested this: if all sales_periods are over active **is not false** and $min_period will be FALSE in these cases.

Reply:
So event though the documentation states that active takes sales periods into consideration it's not doing that for the ticket types sales periods. Hence the event is still active event though there are no active sales periods. In our case active just tells us whether the event is paused/unpaused.

Even though it does create a notice it seems that the ticket-status will be correctly set to closed anyway.

Reply:
Even though the code appears to output the correct status anyway, I don't think it's a good idea to rely on that.

It's probably best to verify that there's in fact any sales periods that are not over before trying to find the min period.